lib/vector-core/buffers/src/buffer_usage_data.rs

```diff
@@ -0,0 +1,87 @@
+use core_common::internal_event::emit;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::{sync::Arc, time::Duration};
+use tokio::time::interval;
+use tracing::{Instrument, Span};
+
+use crate::internal_events::{BufferEventsReceived, BufferEventsSent, EventsDropped};
+
+pub struct BufferUsageData {
+    received_event_count: AtomicUsize,
+    received_event_byte_size: AtomicUsize,
+    sent_event_count: AtomicUsize,
+    sent_event_byte_size: AtomicUsize,
+    dropped_event_count: Option<AtomicUsize>,
+}
+
+impl BufferUsageData {
+    pub fn new(dropped_event_count: Option<AtomicUsize>) -> Self {
+        Self {
+            received_event_count: AtomicUsize::new(0),
+            received_event_byte_size: AtomicUsize::new(0),
+            sent_event_count: AtomicUsize::new(0),
+            sent_event_byte_size: AtomicUsize::new(0),
+            dropped_event_count,
+        }
+    }
+
+    pub fn init_instrumentation(buffer_usage_data: &Arc<BufferUsageData>, span: Span) {
```

Comment:
The split between `init_instrumentation` and `new` is surprising to me. It doesn't seem like you ever call `new` without calling one and exactly one `init_instrumentation`, is that true? Can you help me understand why these are split vs. why we wouldn't pop all of this into `new`?

Reply:
Great point. When I first wrote `BufferUsageData`, I had the `init_instrumentation` logic written out separately in `buffers/lib.rs` and later moved it into `BufferUsageData`. It hadn't occurred to me to put it in `new` since I often associate `new` as just performing basic construction. Moving the logic makes sense to me.